jlm/llvm/opt/CommonNodeElimination.cpp

```diff
@@ -0,0 +1,1044 @@
+      markRegion(*theta.subregion(), context);
+    }
+  }
+
+  // Partition theta outputs
+  std::unordered_map<CongruenceSets::Index, CongruenceSets::Index> resultToOutputSetMapping;
+  for (auto & loopVar : loopVars)
+  {
+    // invariant loop variables become followers of the input origin
+    if (rvsdg::ThetaLoopVarIsInvariant(loopVar))
+    {
+      const auto inputOriginSet = congruenceSets.getSetFor(*loopVar.input->origin());
+      congruenceSets.addFollower(inputOriginSet, *loopVar.output);
+      continue;
+    }
+
+    // Other loop variable outputs are partitioned based on the origins of the post results.
+    const auto resultSet = congruenceSets.getSetFor(*loopVar.post->origin());
+    const auto it = resultToOutputSetMapping.find(resultSet);
+    if (it != resultToOutputSetMapping.end())
+    {
+      congruenceSets.addFollower(it->second, *loopVar.output);
+    }
+    else
+    {
+      const auto outputSet = congruenceSets.getOrCreateSetForLeader(*loopVar.output);
+      resultToOutputSetMapping.emplace(resultSet, outputSet);
+    }
+  }
+}
+
+/**
+ * Marks the given lambda node and its subregion.
+ * Context variables are congruent if their origins are congruent.
+ * All other arguments are given distinct congruence sets.
+ * The lambda node itself becomes its own leader.
+ * @param lambda the lambda node
+ * @param context the current context of the marking phase.
+ */
+static void
+markLambda(const rvsdg::LambdaNode & lambda, CommonNodeElimination::Context & context)
+{
+  markSubregionsFromInputs(lambda, context);
+
+  // A lambda output is always unique
+  markNodeAsLeader(lambda, context);
+}
+
+/**
+ * Marks the given phi node and its subregion.
+ * Context variables are congruent if their origins are congruent.
+ * All other arguments are given distinct congruence sets.
+ * The phi node itself becomes its own leader.
+ * @param phi the phi node
+ * @param context the current context of the marking phase.
+ */
+static void
+markPhi(const rvsdg::PhiNode & phi, CommonNodeElimination::Context & context)
+{
+  markSubregionsFromInputs(phi, context);
+
+  // A phi node is always unique
+  markNodeAsLeader(phi, context);
+}
+
+static void
+markDelta(const rvsdg::DeltaNode & delta, CommonNodeElimination::Context & context)
+{
+  // A delta node is always unique
+  markNodeAsLeader(delta, context);
+}
+
+static void
+markStructuralNode(const rvsdg::StructuralNode & node, CommonNodeElimination::Context & context)
+{
+  rvsdg::MatchTypeOrFail(
+      node,
+      [&](const rvsdg::GammaNode & gamma)
+      {
+        markGamma(gamma, context);
+      },
+      [&](const rvsdg::ThetaNode & theta)
+      {
+        markTheta(theta, context);
+      },
+      [&](const rvsdg::LambdaNode & lambda)
+      {
+        markLambda(lambda, context);
+      },
+      [&](const rvsdg::PhiNode & phi)
+      {
+        markPhi(phi, context);
+      },
+      [&](const rvsdg::DeltaNode & delta)
+      {
+        markDelta(delta, context);
```

Comment:
The markDelta function can be inlined. It is literally a single line of code and comment.